src/ConnectedMode/QualityProfiles/QualityProfileDownloader.cs

```diff
@@ -0,0 +1,202 @@
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.ConnectedMode.Binding;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.Integration;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.QualityProfiles
+{
+    internal interface IQualityProfileDownloader
+    {
+        /// <summary>
+        /// Ensures that the Quality Profiles for all supported languages are to date
+        /// </summary>
+        Task<bool> UpdateAsync(BoundSonarQubeProject boundProject, IProgress<FixedStepsProgress> progress, CancellationToken cancellationToken);
+    }
+
+    internal class QualityProfileDownloader : IQualityProfileDownloader
+    {
+        private readonly IBindingConfigProvider bindingConfigProvider;
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly IConfigurationPersister configurationPersister;
+        private readonly ISolutionBindingOperation solutionBindingOperation;
+
+        private readonly ILogger logger;
+
+        private readonly IEnumerable<Language> languagesToBind;
+
+        [ImportingConstructor]
+        public QualityProfileDownloader(
+            ISonarQubeService sonarQubeService,
+            IBindingConfigProvider bindingConfigProvider,
+            IConfigurationPersister configurationPersister,
+            ISolutionBindingOperation solutionBindingOperation,
+            ILogger logger) :
+            this(
+                sonarQubeService,
+                bindingConfigProvider,
+                configurationPersister,
+                solutionBindingOperation,
+                logger,
+                languagesToBind: Language.KnownLanguages
+                )
+        { }
+
+        internal /* for testing */ QualityProfileDownloader(
+            ISonarQubeService sonarQubeService,
+            IBindingConfigProvider bindingConfigProvider,
+            IConfigurationPersister configurationPersister,
+            ISolutionBindingOperation solutionBindingOperation,
+            ILogger logger,
+            IEnumerable<Language> languagesToBind)
+        {
+            this.bindingConfigProvider = bindingConfigProvider;
+            this.sonarQubeService = sonarQubeService;
+            this.configurationPersister = configurationPersister;
+            this.solutionBindingOperation = solutionBindingOperation;
+            this.logger = logger;
+            this.languagesToBind = languagesToBind;
+        }
+
+        public async Task<bool> UpdateAsync(BoundSonarQubeProject boundProject, IProgress<FixedStepsProgress> progress, CancellationToken cancellationToken)
+        {
+            // TODO - CancellableJobRunner
+            // TODO - threading
+            // TODO - skip downloading up to date QPs
+
+            EnsureProfilesExistForAllSupportedLanguages(boundProject);
```

Comment:
Previously `BindingProcessImpl` looped round list of `languagesToBind` and downloaded a QP for each one -> it would have fetching the QP for any missing languages/partial bindings.

With the benefit of hindsight, I probably didn't need to explicitly add this new `Ensure...` method yet for the same reason. However, I still think it makes sense to have it here as it explicitly sets the QP timestamp to `DateTime.Minimum`, which we'll use when checking for out-of-date profiles later.